File: updater/src/main/java/de/darkatra/patcher/updater/util/ProcessUtils.java
package de.darkatra.patcher.updater.util;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ProcessUtils {

	public static Process run(final Path executable, final String... args) throws IOException {
		return run(executable.toFile(), args);
	}

	public static Process run(final File executable, final String... args) throws IOException {
		return run(executable, executable.getParentFile(), args);
	}

	public static Process run(final File executable, final File workingDirectory, final String... args) throws IOException {

		return run(executable.getAbsolutePath(), workingDirectory, args);
	}

	public static Process runJar(final Path jar, final String... args) throws IOException {
		return run(
			Stream.concat(Stream.of("java", "-jar", jar.toFile().getName()), Stream.of(args)).collect(Collectors.toList()),
			jar.toFile().getParentFile()
		);
	}

	public static Process run(final String command, final File workingDirectory, final String... args) throws IOException {
		return run(
			Stream.concat(Stream.of(command), Stream.of(args)).collect(Collectors.toList()),
			workingDirectory
		);
	}

	public static Process run(final List<String> command, final File workingDirectory) throws IOException {
		final ProcessBuilder processBuilder = new ProcessBuilder(command);
		processBuilder.directory(workingDirectory);
		return processBuilder.start();
	}
}

File: src/jvmMain/kotlin/de/darkatra/bfme2/util/ProcessUtils.kt
package de.darkatra.bfme2.util

import java.nio.file.Path
import kotlin.io.path.name

object ProcessUtils {

    fun runJar(jar: Path, args: Array<String> = emptyArray()): Process {
        return run(
            listOf("java", "-jar", jar.name, *args),
            jar.parent
        )
    }

    fun run(command: List<String>, workingDirectory: Path): Process {
        return ProcessBuilder(command).apply {
            directory(workingDirectory.toFile())
        }.start()
    }
}
